java/compilers/Symbol.java

```diff
@@ -5,26 +5,48 @@
 enum SymbolType { NONTERMINAL, TERMINAL, RULE, ALT, EOF, LAMBDA }
 
 class Symbol {
-    static final Symbol LAMBDA = new Symbol("lambda");
-    static final Symbol EOF = new Symbol("$");
+    static final Symbol LAMBDA = new Symbol("lambda", SymbolType.LAMBDA);
+    static final Symbol EOF = new Symbol("$", SymbolType.EOF);
+    private static final Symbol ALT = new Symbol("|", SymbolType.ALT);
+    private static final Symbol RULE = new Symbol("->", SymbolType.RULE);
 
     String token;
     SymbolType type;
 
-    Symbol(String token) {
+    Symbol() {
+	    token = "";
+	    type = null;
+    }
+    
+    private Symbol(String token, SymbolType type) {
         this.token = token;
-        type = matchSymbol(token);
+        this.type = type;
+    }
+
+    static Symbol of(String token) {
+        switch (token) {
+            case "->": return RULE;
+            case "|": return ALT;
+            case "$": return EOF;
+            case "lambda": return LAMBDA;
+            default:
+                if (token.matches("[a-z]+"))
+                    return new Symbol(token, SymbolType.TERMINAL);
+                if (token.matches("[A-Z]+"))
+                    return new Symbol(token, SymbolType.NONTERMINAL);
+                return new Symbol(token, null);
+        }
     }
 
-    public boolean isTerminal() {
-        return SymbolType.TERMINAL.equals(type);
+    boolean isTerminal() {
+        return type == SymbolType.TERMINAL;
     }
 
-    public boolean isNonTerminal() {
-        return SymbolType.NONTERMINAL.equals(type);
+    boolean isNonTerminal() {
+        return type == SymbolType.NONTERMINAL;
```

Comment:
Mostly for style, to be more consistent.

Reply:
I see that they're all the same now, but that's just because you changed them all from the old style.  I was wondering why you prefer this style specifically.